src/createReducer.ts

```diff
@@ -8,29 +8,31 @@ function composeMiddleware(chain) {
   };
 }
 
-const createReducer = (...middlewares) => (reducer, initialState, initializer = value => value) => {
-  const ref = useRef(initializer(initialState));
-  const [, setState] = useState(ref.current);
-  let middlewareDispatch = (_ = {}) => {
-    throw new Error(
-      'Dispatching while constructing your middleware is not allowed. ' +
-        'Other middleware would not be applied to this dispatch.'
+const createReducer = (...middlewares) => {
+  const composedMiddleware = composeMiddleware(middlewares);
+
+  return (reducer, initialState, initializer = value => value) => {
+    const ref = useRef(initializer(initialState));
+    const [, setState] = useState(ref.current);
+
+    const dispatch = useCallback(action => {
+      ref.current = reducer(ref.current, action);
+      setState(ref.current);
+      return action;
+    }, []);
```

Comment:
I'm not sure if `reducer` should be added to the dependency array? As far as I know it's not common to swap out a reducer with`useReducer`?

Reply:
Yeah you are right. We can swap [reducers](https://codesandbox.io/s/nostalgic-proskuriakova-q85mr) with `useReducer`. So we need add `reducer` to the dependency array. And add

```
  useMemo(() => {
    dispatchRef.current = composedMiddleware(
      {
        getState: () => ref.current,
        dispatch: (...args) => dispatchRef.current(...args),
      },
      dispatch
    );
  }, [dispatch]);
```

because with `useEffect` we have double dispatch initialization [screencast](https://take.ms/l3N88)

What do you think?

Reply:
Would [useUpdateEffect](https://github.com/streamich/react-use/blob/master/docs/useUpdateEffect.md) fix the double dispatch initialization?

Reply:
With `useUpdateEffect` works pretty nice.